packages/SwingSet/src/liveslots/virtualObjectManager.js

```diff
@@ -411,6 +449,27 @@ export function makeVirtualObjectManager(
     }
   }
 
+  function copyMethods(behaviorTemplate) {
```

Comment:
`copyMethods` is given the user's `behavior` object and *returns` a `behaviorTemplate`, right? Can we rename the argument and `obj` to reflect that? We need to defend against the user's `behavior` object doing something tricky (like changing its entries between two calls to `Object.entries`), but I think it can't do that because we only run `Object.entries()` against it once (in `copyMethods`), but these argument names make me think otherwise.

Also, (and maybe this is an @erights question) should we consider `harden`ing the user's `behavior` object when it is first submitted to `defineKind`? We're no longer using their records to build our facets (we build entirely new records), so having them pre-`harden`ed shouldn't interfere with the new objects being made `Far`.

Reply:
> should we consider `harden`ing the user's `behavior` object when it is first submitted to `defineKind`? We're no longer using their records to build our facets (we build entirely new records), so having them pre-`harden`ed shouldn't interfere with the new objects being made `Far`.

That's a good point. Yes, we should preemptively `harden`.

Reply:
I'm plucking the methods out of the `behavior` object that's passed in to guard against other crap that the user might have hung off the object.  After extracting the methods there's really no benefit to `harden`ing (or otherwise fiddling with) the original object.

Reply:
I agree there is no strong reason to `harden` this one. As long as the extraction is careful not to be confusable by a self-mutating behavior object. For example, doing the extraction all at once with an `Object.getOwnPropertyDescriptors` and then fully ignoring the original object would be safe.

Reply:
That's an interesting wrinkle I hadn't thought about: is `Object.getOwnPropertyDescriptors` atomic in a way that iterating over the array returned by `Object.entries` is not?  Thinking about it, that seems right, since if you use `Object.entries`, each pass of the iteration gives the object being iterated over the opportunity to do something funky mid-iteration if one or more of the properties is a getter.  (Of course in order to use the value returned by `Object.getOwnPropertyDescriptors` you have to iterate over _it_ using `Object.entries` or one of its brethren, which is kind of a pain and feels at first glance like you're just recursing on the problem -- but presumably you are now in exclusive possession of _that_ object in a way that you are not in exclusive possession of the original object).  On the other hand, it's unclear to me what mischief somebody could cause by having the behavior object modify itself during iteration, since `defineKind` is only going to iterate over the behavior object once, so any modification is as if that was what it was supposed to do all along.  All `defineKind` cares about is obtaining a set of method functions, so if the object squirmed around somehow while being iterated I don't see why `defineKind` would care. It got its bag of methods from the thing its caller gave it, so where's the harm? Seems to me the worst case is they wrote some weird code that they might not understand later, but anyone can confuse themselves if they work at it.